aptos-move/framework/aptos-framework/sources/AuctionHouse.move

```diff
@@ -0,0 +1,878 @@
+            locked_tokens: Option::some(parse_token_ids(sender, token_creators, token_collection_names, token_names)),
+        })
+    }
+
+    fun is_auction_active(start_time: u64, duration: u64): bool {
+        let current_time = Timestamp::now_microseconds();
+        current_time <= start_time + duration && current_time >= start_time
+    }
+
+    fun is_auction_complete(start_time: u64, duration: u64): bool {
+        let current_time = Timestamp::now_microseconds();
+        current_time > start_time + duration
+    }
+
+    public(script) fun buy_token_script(
+        sender: &signer,
+        seller: address,
+        guid_creation_num: u64
+    ) acquires ListingData {
+        let sender_addr = Signer::address_of(sender);
+        assert!(sender_addr != seller, EINVALID_BUYER);
+
+        let listing_id = GUID::create_id(seller, guid_creation_num);
+        let listing_data = borrow_global<ListingData>(seller);
+        let listed_items = &listing_data.listed_items;
+        let listed_item = Table::borrow(listed_items, listing_id);
+
+        assert!(is_auction_active(listed_item.start_time, listed_item.duration), ELISTING_INACTIVE);
+        let buy_now_price = Option::get_with_default(&listed_item.buy_now_price, 0);
+        assert!(listed_item.min_selling_price == buy_now_price, EBUY_NOW_UNAVAILABLE);
+
+        buy_now(sender, seller, listing_id, buy_now_price);
+    }
+
+    fun buy_now(
+        sender: &signer,
+        seller: address,
+        listing_id: ID,
+        price: u64
+    ) acquires ListingData {
+        let listing_data = borrow_global_mut<ListingData>(seller);
+        let listed_item = Table::borrow_mut(&mut listing_data.listed_items, listing_id);
+
+        Event::emit_event<BuyNowEvent>(
+            &mut listing_data.buy_now_events,
+            BuyNowEvent { id: listing_id, amount: price },
+        );
+
+        let coin = Coin::withdraw<TestCoin>(sender, price);
+        Coin::deposit(seller, coin);
+
+        let locked_tokens = Option::extract(&mut listed_item.locked_tokens);
+        let key = IterableTable::head_key<TokenId, Token>(&locked_tokens);
+        while (Option::is_some(&key)) {
+            let (val, _, next) = IterableTable::remove_iter<TokenId, Token>(&mut locked_tokens, *Option::borrow(&key));
+            Token::deposit_token(sender, val);
+            key = next;
+        };
+
+        IterableTable::destroy_empty(locked_tokens);
+
+        let ListedItem {
+            min_selling_price: _,
+            buy_now_price: _,
+            duration: _,
+            start_time: _,
+            current_bid: _,
+            current_bidder: _,
+            locked_tokens: locked_tokens,
+        } = Table::remove(&mut listing_data.listed_items, listing_id);
+
+        Option::destroy_none(locked_tokens);
+    }
+
+    // places a bid on an active auction
+    // the placed bid should be more than the current highest bid on the auction
+    public(script) fun bid_script(
```

Comment:
can we do it more directly to get rid the claim functions?
the coin and token transfer directly happens within this function.  You can take a look at this https://github.com/aptos-labs/aptos-core/blob/7343c6453fe1ec843f3e1839804b5ed0515b145b/aptos-move/framework/aptos-token/sources/token_coin_swap.move#L73